### DISCUSSION TITLE
LIS DA crashes

### DISCUSSION THREAD

**Discussion by sunnydays99** [NASA-LIS/LISF#1460] [Q&A]
Dear community,

I have been trying to understand why my data assimilation runs keeps on quitting but to no avail. In the retrospective runmode it stops after the first 3 hour and I receive a SIGSEGV: Segmentation fault - invalid memory reference error. Changing the attributes or perturbation files did not help yield to any other output. 
If anyone knows where I should look for to fix the error I would gladly appreciate any advice.
Thank you for your time

The error message follows:

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
Backtrace for this error:
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
Backtrace for this error:
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
Backtrace for this error:
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x14c8e0023960 in ???
#1  0x14c8e0022ac5 in ???
#2  0x14c8dfc4251f in ???
	at ./signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c:0
#0  0x153c06a23960 in ???
#1  0x153c06a22ac5 in ???
#2  0x153c0664251f in ???
	at ./signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c:0
#0  0x153d4bc23960 in ???
#1  0x153d4bc22ac5 in ???
#2  0x153d4b84251f in ???
	at ./signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c:0
#0  0x14f56c423960 in ???
#1  0x14f56c422ac5 in ???
#2  0x14f56c04251f in ???
	at ./signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c:0
#3  0x55e4cde5e2d3 in clsmf25_gettwspred_
	at ../surfacemodels/land/clsm.f2.5/da_tws/clsmf25_gettwspred.F90:55
#4  0x55e4ceb6173c in lsmdagetobspred_
	at ../core/LIS_lsm_FTable.c:1399
#5  0x55e4cdf6303f in __enksgrace_mod_MOD_enksgrace_increments
	at ../dataassim/algorithm/enksgrace/enksgrace_Mod.F90:291
#6  0x55e4ceb5bbe3 in computeincrements_
	at ../core/LIS_dataassim_FTable.c:436
#7  0x55e4cd76766b in __lis_dataassimmod_MOD_lis_dataassim_run
	at ../core/LIS_dataAssimMod.F90:517
#8  0x55e4cea24ddf in __retrospective_runmod_MOD_lis_run_retrospective
	at ../runmodes/retrospective/retrospective_runMod.F90:203
#9  0x55e4ceb6922d in lisrun_
	at ../core/LIS_runmode_FTable.c:202
#10  0x55e4ce2118aa in lisdrv
	at ../offline/lisdrv.F90:70
#11  0x55e4ce2119b4 in main
	at ../offline/lisdrv.F90:57
#3  0x56333af512d3 in clsmf25_gettwspred_
	at ../surfacemodels/land/clsm.f2.5/da_tws/clsmf25_gettwspred.F90:55
#4  0x56333bc5473c in lsmdagetobspred_
	at ../core/LIS_lsm_FTable.c:1399
#5  0x56333b05603f in __enksgrace_mod_MOD_enksgrace_increments
	at ../dataassim/algorithm/enksgrace/enksgrace_Mod.F90:291
#6  0x56333bc4ebe3 in computeincrements_
	at ../core/LIS_dataassim_FTable.c:436
#7  0x56333a85a66b in __lis_dataassimmod_MOD_lis_dataassim_run
	at ../core/LIS_dataAssimMod.F90:517
#8  0x56333bb17ddf in __retrospective_runmod_MOD_lis_run_retrospective
	at ../runmodes/retrospective/retrospective_runMod.F90:203
#9  0x56333bc5c22d in lisrun_
	at ../core/LIS_runmode_FTable.c:202
#10  0x56333b3048aa in lisdrv
	at ../offline/lisdrv.F90:70
#11  0x56333b3049b4 in main
	at ../offline/lisdrv.F90:57
#3  0x5604c10302d3 in clsmf25_gettwspred_
	at ../surfacemodels/land/clsm.f2.5/da_tws/clsmf25_gettwspred.F90:55
#4  0x5604c1d3373c in lsmdagetobspred_
	at ../core/LIS_lsm_FTable.c:1399
#5  0x5604c113503f in __enksgrace_mod_MOD_enksgrace_increments
	at ../dataassim/algorithm/enksgrace/enksgrace_Mod.F90:291
#6  0x5604c1d2dbe3 in computeincrements_
	at ../core/LIS_dataassim_FTable.c:436
#7  0x5604c093966b in __lis_dataassimmod_MOD_lis_dataassim_run
	at ../core/LIS_dataAssimMod.F90:517
#8  0x5604c1bf6ddf in __retrospective_runmod_MOD_lis_run_retrospective
	at ../runmodes/retrospective/retrospective_runMod.F90:203
#9  0x5604c1d3b22d in lisrun_
	at ../core/LIS_runmode_FTable.c:202
#10  0x5604c13e38aa in lisdrv
	at ../offline/lisdrv.F90:70
#11  0x5604c13e39b4 in main
	at ../offline/lisdrv.F90:57
#3  0x5638f69622d3 in clsmf25_gettwspred_
	at ../surfacemodels/land/clsm.f2.5/da_tws/clsmf25_gettwspred.F90:55
#4  0x5638f766573c in lsmdagetobspred_
	at ../core/LIS_lsm_FTable.c:1399
#5  0x5638f6a6703f in __enksgrace_mod_MOD_enksgrace_increments
	at ../dataassim/algorithm/enksgrace/enksgrace_Mod.F90:291
#6  0x5638f765fbe3 in computeincrements_
	at ../core/LIS_dataassim_FTable.c:436
#7  0x5638f626b66b in __lis_dataassimmod_MOD_lis_dataassim_run
	at ../core/LIS_dataAssimMod.F90:517
#8  0x5638f7528ddf in __retrospective_runmod_MOD_lis_run_retrospective
	at ../runmodes/retrospective/retrospective_runMod.F90:203
#9  0x5638f766d22d in lisrun_
	at ../core/LIS_runmode_FTable.c:202
#10  0x5638f6d158aa in lisdrv
	at ../offline/lisdrv.F90:70
#11  0x5638f6d159b4 in main
	at ../offline/lisdrv.F90:57
 
[lis.config_clsm_graceda.txt](https://github.com/NASA-LIS/LISF/files/13479036/lis.config_clsm_graceda.txt)
[forcing_pert_attribs.txt](https://github.com/NASA-LIS/LISF/files/13479100/forcing_pert_attribs.txt)
[forcing_attribs.txt](https://github.com/NASA-LIS/LISF/files/13479101/forcing_attribs.txt)
[clsm_tws_pertattribs.txt](https://github.com/NASA-LIS/LISF/files/13479102/clsm_tws_pertattribs.txt)
[clsm_tws_attribs.txt](https://github.com/NASA-LIS/LISF/files/13479104/clsm_tws_attribs.txt)
[grace_tws_pertattribs.txt](https://github.com/NASA-LIS/LISF/files/13479107/grace_tws_pertattribs.txt)
[grace_tws_attribs.txt](https://github.com/NASA-LIS/LISF/files/13479108/grace_tws_attribs.txt)

**Comment by dmocko** [comment]
Hi @sunnydays99 

Are there any further messages that might be helpful at the end of any of your lislog files?

Can you diagnose if you have enough memory available on your machine to run this configuration?

**Comment by sunnydays99** [comment]
I have around 75GB of storage left and the error persists if I try reduce the interval to a month instead of a full year

Here is the one for the log file, with the exception of a Error in nf90_get_att: GRIDTYPE warning nothing else shows up.
[log_graceda.0000.txt](https://github.com/NASA-LIS/LISF/files/13480250/log_graceda.0000.txt)

**Reply by dmocko** [reply]
Thanks - but I mean memory, not available disk space storage.

Our job queue system gives a report of the amount of memory used during the simulation, e.g.:

```
  Max Real Mem Used (on any node) : 41.68G
  Max Pct Memory Used             : 3.74%
```

**Reply by sunnydays99** [reply]
Okay I see, I am running LIS within my own virtual machine, not sure how to check for the specific memory usage outside from what the system monitor reports. In total my VM has 40GB of memory and during the run around each LIS executable uses around  2GB of virtual memory and 200MB of actual memory storage (less then 7% of the available amount).

**Comment by dmocko** [comment]
Hi @TimLahmers @KathyNie 

Can you please help try to figure out this issue?

@sunnydays99 - what branch are you using?

**Reply by TimLahmers** [reply]
@sunnydays99 I have done some work with GRACE DA, but @KathyNie is more familiar with this LIS component. 

Before we go into this further, could you please confirm if the model runs with an open-loop configuration (i.e., your domain runs with identical settings but without DA)? That will help us confirm that there isn't a problem with your LSM, forcing, or other LIS settings.

@dmocko

**Reply by sunnydays99** [reply]
Hello everyone,
I am working with the public branch version 7.4. 
My OL run successfully generated outputs for the whole study period and afterwards then other LDT runs based on the open-loop output run without issues as well (Anomaly correction and Ensemble Restart Generation). I also checked whether the stack size is causing the issues but that is unlimited. 
Kind regards

**Reply by KathyNie** [reply]
@sunnydays99 glad to help. Could you share your ldt.config and lis.config files?

**Reply by sunnydays99** [reply]
Yes absolutely!

here are the general files I used to compile LDT and LIS:
[configure.lis.txt](https://github.com/NASA-LIS/LISF/files/13532242/configure.lis.txt)
[configure.ldt.txt](https://github.com/NASA-LIS/LISF/files/13532244/configure.ldt.txt)


and these are the config files for the parameter estimation, open loop run, ensemble restart generation and anomaly correction. Kind regards
[lis.config_clsm.ol.txt](https://github.com/NASA-LIS/LISF/files/13532247/lis.config_clsm.ol.txt)
[ldt.config.txt](https://github.com/NASA-LIS/LISF/files/13532248/ldt.config.txt)
[ldt.config.clsm_params.txt](https://github.com/NASA-LIS/LISF/files/13532249/ldt.config.clsm_params.txt)
[ldt.config.clsm_ac.txt](https://github.com/NASA-LIS/LISF/files/13532250/ldt.config.clsm_ac.txt)

**Reply by KathyNie** [reply]
Hi @sunnydays99, I noticed one error in your lis.config_clsm_graceda.txt file, the running mode should be switched to "ensemble smoother" instead of "retrospective". Could you try this fix and see if that helps fix the issue? Besides, I would also suggest turn on the perturbation bias correction, which corresponds to the following setting:

Apply perturbation bias correction:        1

Hope that helps and please keep us updated.

**Reply by sunnydays99** [reply]
@KathyNie thank you for your message! I have adjusted the configuration file to use the ensemble smoother mode instead as well as turned on the perturbation bias correction. Now I get surface model outputs for the first day followed by a cryptic error message follows with "error in the return code prgm stopping" I tried changing the LIS time window parameter from 1 da to 1 mo and 1yr but the same error message shows up. 
I will attach the lislog below: 
[log_graceda.0000.txt](https://github.com/NASA-LIS/LISF/files/13572273/log_graceda.0000.txt)

**Reply by KathyNie** [reply]
@sunnydays99 How about the last few lines in the remaining log files? Are they the same as the one you attached? Besides, could you share the full info of the error message associated with the crash? For instance, if you are running under an interactive node, you should be able to see that directly on the terminal window when it crashed. 

And maybe also test the following: 1) comment out the entry of "LIS time window interval"; 2) for the "Surface model output interval", change "3hr" to "1da" and see how it goes.

**Reply by sunnydays99** [reply]
@KathyNie 
I followed your advice on commenting out the LIS time window interval, which doesn't appear to have an impact on the run itself. Changing the surface model output interval from "3hr" to "1da" yields the program stopping at the same 1day threshold. All log files are the same like the one provided with the exception that log003 last line ends at LIS cycle 23.45. I have also been testing the perturbation frequencies and if I set the state perturbation frequency at 30mn, 1h, 2h, 6h, 12hr, the same error message shows up and the run gets aborted. Could the problem be related to this?
Kind regards

The error message that shows up in the terminal is following: 

Error in the return code, Prgm Stopping...
Abort(1) on node 3 (rank 3 in comm 496): application called MPI_Abort(comm=0x84000001, 1) - process 3
 Error in the return code, Prgm Stopping...
Abort(1) on node 1 (rank 1 in comm 496): application called MPI_Abort(comm=0x84000001, 1) - process 1
 Error in the return code, Prgm Stopping...
Abort(1) on node 2 (rank 2 in comm 496): application called MPI_Abort(comm=0x84000001, 1) - process 2
 Error in the return code, Prgm Stopping...
Abort(1) on node 0 (rank 0 in comm 496): application called MPI_Abort(comm=0x84000002, 1) - process 0

**Reply by KathyNie** [reply]
@sunnydays99 Thanks for testing those options. In your output directory, is the directories named "DAOBS", "DAPERT", "EnKF" being created besides the "log" and "SURFACEMODEL" directories?

**Reply by sunnydays99** [reply]
@KathyNie These are not begin created, as the run stops beforehand. 
But when I changed the Data assimilation output interval for diagnostics: from 1da to 12hr and the same for the perturbation restart interval , I managed to output the both directories EnKF and DAPERT after which the same error in the return code message showed up and the run aborted. I tried to use ncview to visualize the ensemble spread and it appear as if all 3 are zeros.
I will attach the EnKF output below, maybe that's the culprit?
[LIS_DA_EnKF_200301011200_spread.a01.d01.nc.txt](https://github.com/NASA-LIS/LISF/files/13591521/LIS_DA_EnKF_200301011200_spread.a01.d01.nc.txt)

**Reply by KathyNie** [reply]
Interesting! I'm not sure if that is the cause, but let's try not output the ensemble spread and see if that helps pass this crash point:

"Data assimilation output ensemble spread:            0"

**Reply by sunnydays99** [reply]
Okay I see, I tried a run without outputting the ensemble spread without any change to the crash point unfortunately

**Reply by KathyNie** [reply]
Ok, I will run your case on my side and take a close look. Will get back to you if I have any questions or if I find any bugs.

**Reply by KathyNie** [reply]
@sunnydays99 Could you also share those files in your INPUT directory such as the CLSM related parameter files?

**Comment by sunnydays99** [comment]
thank  you so much for taking a closer look at the error message.The clsm parameter files is as follows : 
[Uploading lis_input.gdas.clsm.nc.txt…]()

**Reply by KathyNie** [reply]
Hi, it seems that it is not fully uploaded. Would you mind upload it again?

**Reply by KathyNie** [reply]
Hi @sunnydays99 I'm also curious to learn more about your research question regarding CLSM-based GRACE-DA. Additionally it would be helpful if you could identify your organization. That would help us have a better context on how we can help. Currently our GRACE-DA practices are primarily centered around the Noah-MP land surface model and I just learned that when working with CLSM, you need to contact the corresponding team to get a set of parameter specifically to your configuration requirements. If those parameters are not configured correctly, it could potentially impact the capability with GRACE-DA.

**Reply by sunnydays99** [reply]
@KathyNie thank you for your message. I tried to set up GRACE-DA for Noah-MP instead and that is running without issues. It might have been the missing parameters for CLSM that were impacting the performance.
